### PR TITLE
feat(react): remove unnecessary dependencies from @nrwl/react

### PR DESCRIPTION
**Pull request by jaysoo**
This PR removes packages such as `@nrwl/webpack` and `@nrwl/cypress` as dependencies from `@nrwl/react`. The goal is to avoid installing things that are not needed. For example, if vite is used then webpack should not be installed. Same thing for vitest replacing jest.

## Technical details

To make sure we can still use other generators inside any generator (such as `webpackInitGenerator` from `@nrwl/webpack`), a new utility function is added to `@nrwl/devkit`: `ensurePackage`.

This function takes a package and an expected version, and checks to make sure it is already installed in the workspace.

```ts
export async function ensurePackage(
  tree: Tree,
  pkg: string,
  requiredVersion: string,
  options: {
    cwd?: string;
    dev?: boolean;
    packageJsonPath?: string;
    packageManager?: 'npm' | 'yarn' | 'pnpm';
    throwOnMissing?: boolean;
  } = {}
): Promise<void> {
  //...
}
```

If the package is not found, then one of two things happens:

1. If user passed `--dryRun`, then `throwOnMissing` is set to `true` and an error is thrown with a message like `Required package <pkg>@<version> is missing. Run npm install -D <pkg>@<version>, and then try again.`
2. If user did _not_ pass `--dryRun` then the missing package is installed before continuing with the rest of the invoking generator. This will be the common case, for example when running `create-nx-workspace`.

## Results / Comparisons

A normal React+Vite workspace using `--preset=react-standalone` will result in **365MB** of `node_modules`  prior to this change. After this change we end up with **259MB** of `node_modules`, or a **106MB** reduction in size.

A plain vite project created via `npm create vite` (select React and TypeScript), then adding the following packages:

```shell
npm i prettier \
react-test-renderer \
vitest \
@vitest/ui \
@vitejs/plugin-react \
@types/node \
@testing-library/react \
vite-tsconfig-paths
```

In that plain vite setup, `node_modules` is **151MB** in size. There are still some improvements we can make (such as removing `rxjs`), but that is out of scope for this PR.

If you add the eslint plugins, then the vite project goes to **208MB** in size: 

```shell
npm i eslint \
eslint-config-prettier \
eslint-plugin-import \
eslint-plugin-jsx-a11y \
eslint-plugin-react \
eslint-plugin-react-hooks \
eslint-config-prettier \
eslint-plugin-import \
eslint-plugin-jsx-a11y \
eslint-plugin-react \
eslint-plugin-react-hooks \
vite-plugin-eslint \
@typescript-eslint/eslint-plugin \
@typescript-eslint/parser
```

> Note: There seems to be duplicated packages at different versions such as `aria-query` between eslint plugins and testing library.

If you bring in the nx-specific packages, that's another **39MB** for a total **247MB** (or a 12 MB difference from our `react-standalone` preset):

```
 6.0M	node_modules/nrwl
 4.2M	node_modules/nx
 12M	node_modules/@parcel
 17M	node_modules/rxjs
```

## Other notes

* `@nrwl/react` now relies fully on vite/webpack/jest/cypress init generators to install the necessary packages.
* Added `uiFramework` option to webpack init generator, same option already supported in cypress and storybook.
* The `AsyncIterable` helpers from `@nrwl/js` have been moved to `@nrwl/devkit` so we can remove the JS dependency from `@nrwl/react`.

### CODE REVIEW

File: packages/devkit/src/utils/package-json.ts [hunk 1]
```diff
@@ -274,3 +277,68 @@ function requiresRemovingOfPackages(
...
+ * @param requiredVersion the version to check
+ * @param {EnsurePackageOptions} options
+ * @returns {Promise<void>}
+ */
+export async function ensurePackage(
+  tree: Tree,
+  pkg: string,
+  requiredVersion: string,
+  options: {
+    cwd?: string;
```
Comment (vsavkin) [inline packages/devkit/src/utils/package-json.ts:313]: When would it be no workspaceRoot?

File: packages/devkit/src/utils/package-json.ts [hunk 2]
```diff
@@ -274,3 +277,68 @@ function requiresRemovingOfPackages(
...
+    dev?: boolean;
+    packageJsonPath?: string;
```
Comment (vsavkin) [inline packages/devkit/src/utils/package-json.ts:315]: isn't it always package.json at the root?

File: packages/devkit/src/utils/package-json.ts [hunk 3]
```diff
@@ -274,3 +277,68 @@ function requiresRemovingOfPackages(
...
+    packageManager?: 'npm' | 'yarn' | 'pnpm';
```
Comment (vsavkin) [inline packages/devkit/src/utils/package-json.ts:316]: Should we pass it or should we detect it like we do other places?

**Review by vsavkin — APPROVED** [review]

File: packages/devkit/src/utils/package-json.ts (hunk 2 above)
Comment (jaysoo) [reply packages/devkit/src/utils/package-json.ts:315]: Yeah that's true, i'll remove it. Was originally thinking about nested `package.json`, but that doesn't matter anyway for plugins.

File: packages/devkit/src/utils/package-json.ts (hunk 3 above)
Comment (jaysoo) [reply packages/devkit/src/utils/package-json.ts:316]: It's detected, passed for unit tests, but maybe it doesn't matter.

File: docs/generated/devkit/index.md [hunk 4]
```diff
@@ -2114,6 +2221,31 @@ of comments with a replaceCharacter
 
 ---
 
+### tapAsyncIterable
```
Comment (FrozenPandaz) [inline docs/generated/devkit/index.md:2224]: I don't think we should export these.

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]

File: packages/esbuild/src/executors/esbuild/esbuild.impl.ts [hunk 5]
```diff
@@ -15,7 +15,7 @@ import { normalizeOptions } from './lib/normalize';
 
 import { EsBuildExecutorOptions } from './schema';
 import { removeSync, writeJsonSync } from 'fs-extra';
-import { createAsyncIterable } from '@nrwl/js/src/utils/async-iterable/create-async-iterable';
+import { createAsyncIterable } from '@nrwl/devkit';
```
Comment (FrozenPandaz) [inline packages/esbuild/src/executors/esbuild/esbuild.impl.ts:18]: Can you keep this as a deep import?

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]

**Review by FrozenPandaz — APPROVED** [review]

File: packages/react/src/generators/init/init.ts [hunk 6]
```diff
@@ -66,22 +67,54 @@ function updateDependencies(host: Tree, schema: InitSchema) {
   });
 }
 
+function initRootBabelConfig(tree: Tree, schema: InitSchema) {
```
[the remark this answers is not in the document]
Comment (jaysoo) [reply packages/react/src/generators/init/init.ts:70]: I don't see it in master, can be addresses later.

**Comment by ViktorPontinen** [comment]
> I am seeing some odd behavior in [`15.3.0`](https://github.com/nrwl/nx/releases/tag/15.3.0) and I am pretty sure it is related to these changes.
> 
> after upgrading, I needed to add the following modules in order to get `npx nx affected --target=test --parallel=3 --ci --code-coverage` and `npx nx affected --target=build --parallel=3` to work like in `15.2.4`:
> 
> * `@babel/preset-react`
> * `@pmmmwh/react-refresh-webpack-plugin`
> * `@svgr/webpack`
> * `url-loader`
> * ( maybe `react-refresh` ? )
> 
> I will file a bug report later; just wanted to give a quick heads up in case it helps.

Also noticed issues and had to add `react-refresh` as a separate dependency.
Not sure how common our setup is tho so might not affect many.
We run react-scripts (CRA(v5)) via craco (v7) using `nx:run-commands` with `cwd` inside each application.

**Comment by enchorb** [comment]
Same issue here had to `yarn add -D react-refresh`. @FrozenPandaz will this be needed going forward, or a regression that will be reverted in a later version?

**Pull request merged**